File: sbe/binance/spot_3_2/generated-php/spot_sbe/BookTickerSymbolResponse.php
<?php
/**
 * Generated SBE (Simple Binary Encoding) message codec.
 */

class BookTickerSymbolResponse
{
    public const TEMPLATE_ID = 211;
    public const SCHEMA_ID = 3;
    public const SCHEMA_VERSION = 2;
    public const BLOCK_LENGTH = 34;

    public int|float|array|null $priceExponent = null;
    public int|float|array|null $qtyExponent = null;
    public int|float|array|null $bidPrice = null;
    public int|float|array|null $bidQty = null;
    public int|float|array|null $askPrice = null;
    public int|float|array|null $askQty = null;
    public string $symbol = '';

    public function encode(): string
    {
        $buffer = '';

        if ($this->priceExponent !== null) {
            $buffer .= pack('c', $this->priceExponent);
        }
        if ($this->qtyExponent !== null) {
            $buffer .= pack('c', $this->qtyExponent);
        }
        if ($this->bidPrice !== null) {
            $buffer .= pack('q', $this->bidPrice);
        }
        if ($this->bidQty !== null) {
            $buffer .= pack('q', $this->bidQty);
        }
        if ($this->askPrice !== null) {
            $buffer .= pack('q', $this->askPrice);
        }
        if ($this->askQty !== null) {
            $buffer .= pack('q', $this->askQty);
        }

        return $buffer;
    }

    public function decode(string $data): void
    {
        $offset = 0;

        $this->priceExponent = unpack('c', substr($data, $offset, 1))[1];
        $offset += 1;
        $this->qtyExponent = unpack('c', substr($data, $offset, 1))[1];
        $offset += 1;
        $this->bidPrice = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->bidQty = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->askPrice = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->askQty = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
    }
}

File: sbe/binance/spot_3_2/generated-python/spot_sbe/BookTickerSymbolResponse.py
"""Generated SBE (Simple Binary Encoding) message codec."""

import struct
from typing import List, Optional, Tuple, Set
from io import BytesIO

class BookTickerSymbolResponse:
    """SBE message: BookTickerSymbolResponse."""

    TEMPLATE_ID = 211
    SCHEMA_ID = 3
    SCHEMA_VERSION = 2
    BLOCK_LENGTH = 34

    def __init__(self):
        self.price_exponent = None
        self.qty_exponent = None
        self.bid_price = None
        self.bid_qty = None
        self.ask_price = None
        self.ask_qty = None
        self.symbol = b''

    def _decode_var_data(self, data: bytes, offset: int) -> Tuple[bytes, int]:
        """Decode variable-length binary data."""
        pos = offset
        length = struct.unpack_from('<I', data, pos)[0]
        pos += 4
        value = data[pos:pos+length]
        pos += length
        return (value, pos)

    def encode(self) -> bytes:
        """Encode the message to bytes."""
        buffer = BytesIO()

        if self.price_exponent is not None:
            buffer.write(struct.pack('<b', self.price_exponent))
        if self.qty_exponent is not None:
            buffer.write(struct.pack('<b', self.qty_exponent))
        if self.bid_price is not None:
            buffer.write(struct.pack('<q', self.bid_price))
        if self.bid_qty is not None:
            buffer.write(struct.pack('<q', self.bid_qty))
        if self.ask_price is not None:
            buffer.write(struct.pack('<q', self.ask_price))
        if self.ask_qty is not None:
            buffer.write(struct.pack('<q', self.ask_qty))

        return buffer.getvalue()

    def decode(self, data: bytes) -> None:
        """Decode the message from bytes."""
        pos = 0

        self.price_exponent = struct.unpack_from('<b', data, pos)[0]
        pos += 1
        self.qty_exponent = struct.unpack_from('<b', data, pos)[0]
        pos += 1
        self.bid_price = struct.unpack_from('<q', data, pos)[0]
        pos += 8
        self.bid_qty = struct.unpack_from('<q', data, pos)[0]
        pos += 8
        self.ask_price = struct.unpack_from('<q', data, pos)[0]
        pos += 8
        self.ask_qty = struct.unpack_from('<q', data, pos)[0]
        pos += 8

        # Skip to end of block for forward compatibility
        pos = 34


        self.symbol, pos = self._decode_var_data(data, pos)
